File: c++/83. Remove Duplicates from Sorted List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *p = head, *pre = nullptr, *t;
        while (p) {
            if (pre && pre->val == p->val) {
                t = p;
                pre->next = p->next;
                p = p->next;
                delete t;
            } else {
                pre = p;
                p = p->next;
            }
        }
        
        return head;
    }
};

File: python/83. Remove Duplicates from Sorted List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        p = dummy
        while head:
            p.next = head
            p = p.next
            tmp = head.next
            while tmp and tmp.val==head.val:
                tmp = tmp.next
            head = tmp

        p.next = None
        return dummy.next